test/RemoveFiles.test.js

```diff
@@ -34,107 +34,93 @@ describe("RemovedFiles", () => {
 		return;
 	}
 
-	jest.setTimeout(20000);
+	jest.setTimeout(10000);
 
 	function cleanup() {
 		rimraf.sync(tempFolderPath);
 	}
 
-	beforeAll(() => {
+	beforeEach(() => {
 		cleanup();
 		fs.mkdirSync(tempFolderPath);
+
+		// Set file timestamps to 5 seconds earlier,
+		// otherwise the newly-created files will trigger the webpack watch mode to re-compile.
+		let fakeTime = new Date(Date.now() - 5000);
 		fs.writeFileSync(
 			tempFilePath,
 			"module.exports = function temp() {return 'temp file';};\n require('./temp-file2')",
 			"utf-8"
 		);
+		fs.utimesSync(tempFilePath, fakeTime, fakeTime);
 		fs.writeFileSync(
 			tempFile2Path,
 			"module.exports = function temp2() {return 'temp file 2';};",
 			"utf-8"
 		);
+		fs.utimesSync(tempFile2Path, fakeTime, fakeTime);
+		// Wait 1s, otherwise the newly-created files will trigger our test compilers to re-compile.
 	});
-	afterAll(done => {
+	afterEach(() => {
 		cleanup();
-		done();
 	});
 
 	it("should track removed files when they've been deleted in watchRun", done => {
 		const compiler = createSingleCompiler();
 		let watcher;
-		function handleError(err) {
-			if (err) done(err);
-		}
-		setTimeout(() => {
-			fs.unlinkSync(tempFilePath, handleError);
-		}, 2000);
-		compiler.hooks.watchRun.tap("RemovedFilesTest", (compiler, err) => {
-			if (err) {
-				done(err);
-			}
+
+		compiler.hooks.watchRun.tap("RemovedFilesTest", compiler => {
 			const removals = Array.from(compiler.removedFiles);
 			if (removals.length > 0) {
-				setTimeout(() => {
-					expect(removals).toContain(tempFilePath);
-					watcher.close();
-					done();
-				}, 100);
+				expect(removals).toContain(tempFilePath);
+				watcher.close();
+				done();
 			}
 		});
 
-		watcher = compiler.watch(
-			{
-				aggregateTimeout: 50
-			},
-			(err, stats) => {}
-		);
+		watcher = compiler.watch({ aggregateTimeout: 50 }, () => {});
+
+		setTimeout(() => {
+			fs.unlinkSync(tempFilePath);
+		}, 1000);
 	});
 
-	it("should not track removed files when they have not been deleted in watchRun", done => {
+	it("should not track removed files during initial watchRun", done => {
 		const compiler = createSingleCompiler();
+		let watchRunFinished;
 		let watcher;
-		compiler.hooks.watchRun.tap("RemovedFilesTest", (compiler, err) => {
-			if (err) {
-				done(err);
-			}
-			expect(Array.from(compiler.removedFiles)).toHaveLength(0);
-			done();
-			watcher.close();
+		compiler.hooks.watchRun.tap("RemovedFilesTest", compiler => {
+			const removals = Array.from(compiler.removedFiles);
+			watchRunFinished = new Promise(resolve => {
```

Comment:
That looks suspicious. Should it be more like this:

``` js
const watchRunFinished = new Promise(resolve => {
  compiler.hooks.watchRun.tap("RemovedFilesTest", compiler => {
    const removals = Array.from(compiler.removedFiles);
    expect(removals).toHaveLength(0);
    resolve();
  });
});
```

Reply:
But we even bothering with a Promise, the original code should work fine here and is simpler.

Reply:
so one of the issues I ran into was that the initial watchRun hook was being called before `watcher` was assigned, and it would throw an error at `watcher.close()`. In fact, if you look at the original code, `done()` is being called before `watcher.close()`, and I realized that it pretty much just hides the error, because the test succeed before the error can be thrown

Reply:
Ok that makes sense.

This first comment still makes sense, because you should not assume that watchRun is called synchrously in `.watch`.

Maybe it even makes sense to wait for compilation completion before ending the test. `close()` has a callback that is called when the watcher has closed: `watch.close(done)`.